### PR TITLE
WIP:Refactor Idris packaging infrastructure

### PR DESCRIPTION
**Pull request by mpickering**
The main two changes are

1. Completely rewrite how with-packages works to remove use of envHooks
2. The package description is now an idris specific set rather than
    being a subset of the arguments to mkDerivation. This mirrors the
    way Haskell packages are treated.

Possible reviewers: @shlevy @yurrriq @peti 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by yurrriq** [comment]
```
λ darwin-rebuild build
building the system configuration...
these derivations will be built:
  /nix/store/07lca737j409pbk73wivnlqfhnw5g8f5-system-applications.drv
  /nix/store/38m8pfr4d56qrniqmg52bnl5vj0r1jhl-system-path.drv
  /nix/store/gdc4kxck99aywqrckcn7p9hrapq9h92a-darwin-system-17.09pre-git.drv
building path(s) ‘/nix/store/41qsrkc3zpfpy54mfxjyckzjjgfnyag0-system-applications’
building path(s) ‘/nix/store/zq2l39b3wcw52qnd5h7k1qw5kvkg5v2k-system-path’
Use of uninitialized value $propagated in split at /nix/store/9g4wc31j7a2xp22xpgwr0qssfxahxdzl-builder.pl line 146.
collision between `/nix/store/8w7r2nzdk13lh7jbncg4bhfz5lbp8jzz-difftodo-0.2.0/lib/links/libHSasync-2.1.1.1-8yywY4inVGRLJSCg60gBXj-ghc8.0.2.dylib' and `/nix/store/gwxnapsp50wj92y6i9ix6h5w3rpbxk47-idris-1.1.1-with-packages/lib/links/libHSasync-2.1.1.1-8yywY4inVGRLJSCg60gBXj-ghc8.0.2.dylib'
collision between `/nix/store/8w7r2nzdk13lh7jbncg4bhfz5lbp8jzz-difftodo-0.2.0/lib/links/libHSasync-2.1.1.1-8yywY4inVGRLJSCg60gBXj-ghc8.0.2.dylib' and `/nix/store/gwxnapsp50wj92y6i9ix6h5w3rpbxk47-idris-1.1.1-with-packages/lib/links/libHSasync-2.1.1.1-8yywY4inVGRLJSCg60gBXj-ghc8.0.2.dylib'
builder for ‘/nix/store/07lca737j409pbk73wivnlqfhnw5g8f5-system-applications.drv’ failed with exit code 255
cannot build derivation ‘/nix/store/gdc4kxck99aywqrckcn7p9hrapq9h92a-darwin-system-17.09pre-git.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/gdc4kxck99aywqrckcn7p9hrapq9h92a-darwin-system-17.09pre-git.drv’ failed
```

**Comment by mpickering** [comment]
I can't work out which exact derivation you are trying to build there?

**Comment by shlevy** [comment]
Let me know when it's no longer WIP and I'll take a look!

**Comment by mpickering** [comment]
@shlevy It should all work as expected. At least a comment on the structure would be useful. I marked it as WIP to indicate that I would like review from someone with more experience than myself.

File: pkgs/development/idris-modules/build-idris-package.nix [hunk 1]
```diff
@@ -2,40 +2,56 @@
...
+  , version
+  , src
+  , meta
+  , extraBuildInputs ? []
+  , postPatch ? ""
+  , postUnpack ? ""
+  , doCheck ? true
+  }:
+let
+  idris-with-packages = idrisPackages.with-packages (idrisDeps);
```
Comment (shlevy) [inline pkgs/development/idris-modules/build-idris-package.nix]: Drop parens

File: pkgs/development/idris-modules/build-builtin-package.nix [hunk 2]
```diff
@@ -6,19 +6,20 @@ let
   inherit (builtins.parseDrvName idris.name) version;
 in
 build-idris-package {
-  name = "${name}-${version}";
 
-  propagatedBuildInputs = deps;
+
+  pkgName = name;
+  version = version;
```
Comment (shlevy) [inline pkgs/development/idris-modules/build-builtin-package.nix]: `inherit version`

File: pkgs/development/idris-modules/build-builtin-package.nix [hunk 3]
```diff
@@ -6,19 +6,20 @@ let
...
+  idrisDeps = deps;
+
   postUnpack = ''
     sourceRoot=$sourceRoot/libs/${name}
   '';
 
-  postPatch = ''
-    sed -i ${name}.ipkg -e "/^opts/ s|-i \\.\\./|-i $IDRIS_LIBRARY_PATH/|g"
-  '';
+
```
Comment (shlevy) [inline pkgs/development/idris-modules/build-builtin-package.nix]: Superfluous white space

File: pkgs/development/idris-modules/build-idris-package.nix [hunk 4]
```diff
@@ -2,40 +2,56 @@
...
+
+  postUnpack = postUnpack;
+
+
+  # Some packages use the style
+  # opts = -i ../../path/to/package
+  # rather than the declarative pkgs attribute so we have to rewrite the path.
+  postPatch = ''
+    sed -i *.ipkg -e "/^opts/ s|-i \\.\\./|-i ${idris-with-packages}/libs/|g"
+    cat *.ipkg
```
Comment (shlevy) [inline pkgs/development/idris-modules/build-idris-package.nix]: Debug leftover?

File: pkgs/development/idris-modules/build-idris-package.nix [hunk 5]
```diff
@@ -2,40 +2,56 @@
...
-
-    envHooks+=(addIdrisLibs)
+{ stdenv, idrisPackages, gmp }:
+  { idrisDeps ? []
+  , pkgName
+  , version
+  , src
+  , meta
+  , extraBuildInputs ? []
+  , postPatch ? ""
```
Comment (shlevy) [inline pkgs/development/idris-modules/build-idris-package.nix]: Looks like this isn't referenced below.

File: pkgs/development/idris-modules/build-idris-package.nix [hunk 6]
```diff
@@ -2,40 +2,56 @@
...
 
+  src = src;
+
   buildPhase = ''
-    ${idris}/bin/idris --build *.ipkg
+    ${idris-with-packages}/bin/idris --build *.ipkg
   '';
 
-  doCheck = true;
+  doCheck = doCheck;
```
Comment (shlevy) [inline pkgs/development/idris-modules/build-idris-package.nix]: `inherit doCheck`, you can have multiple `inherits` at once (e.g. `inherit postUnpack doCheck;`)

File: pkgs/development/idris-modules/default.nix [hunk 7]
```diff
@@ -25,22 +25,35 @@
...
     };
 
-    files = builtins.filter (n: n != "default") (pkgs.lib.mapAttrsToList (name: type: let
-      m = builtins.match "(.*)\\.nix" name;
-    in if m == null then "default" else builtins.head m) (builtins.readDir ./.));
-  in (builtins.listToAttrs (map (name: {
-    inherit name;
-
-    value = callPackage (./. + "/${name}.nix") {};
-  }) files)) // {
```
Comment (shlevy) [inline pkgs/development/idris-modules/default.nix]: We can certainly clean this up/document it more, but :-1: for me on removing the autodiscovery.

File: pkgs/development/idris-modules/with-packages.nix [hunk 8]
```diff
@@ -1,46 +1,20 @@
...
+  buildInputs = [ makeWrapper ];
 
-  unpackPhase = ''
-    cat >idris.c <<EOF
-    #include <stdlib.h>
-    #include <unistd.h>
-    #include <stdio.h>
+  postBuild = ''
+    wrapProgram $out/bin/idris \
+      --set IDRIS_LIBRARY_PATH $out/libs
```
Comment (shlevy) [inline pkgs/development/idris-modules/with-packages.nix]: Won't this overwrite a user-set `IDRIS_LIBRARY_PATH`?

**Review by shlevy — CHANGES_REQUESTED** [review]
Overall looks OK, see comments in line.

File: pkgs/development/idris-modules/default.nix (hunk 7 above)
Comment (mpickering) [reply pkgs/development/idris-modules/default.nix]: I don't like this style at all, it is too implicit and relied on a false assumption that you want an attribute for every `*.nix` file in a repo. There are not many places in nixpkgs which I could find using it.

File: pkgs/development/idris-modules/with-packages.nix (hunk 8 above)
Comment (mpickering) [reply pkgs/development/idris-modules/with-packages.nix]: That's a good point. Would a good solution to be to also symlink the already set `IDRIS_LIBRARY_PATH` into the right place?

File: pkgs/development/idris-modules/build-idris-package.nix (hunk 6 above)
Comment (mpickering) [reply pkgs/development/idris-modules/build-idris-package.nix]: I also don't like the inherit style but I will admit that it is very commonly used so I will change that.

File: pkgs/development/idris-modules/with-packages.nix (hunk 8 above)
Comment (shlevy) [reply pkgs/development/idris-modules/with-packages.nix]: Originally I just had it respect the original `IDRIS_LIBRARY_PATH` and not override it. The real solution is to let IDRIS_LIBRARY_PATH have multiple colon-separated entries though.

Comment (mpickering) [reply pkgs/development/idris-modules/with-packages.nix]: I don't think it is terrible to overwrite the `IDRIS_LIBRARY_PATH` for now, users can easily package their libraries and let nix manage them.

File: pkgs/development/idris-modules/build-idris-package.nix (hunk 4 above)
Comment (mpickering) [reply pkgs/development/idris-modules/build-idris-package.nix]: No, this is necessary to build the base packages.

Comment (shlevy) [reply pkgs/development/idris-modules/build-idris-package.nix]: The `cat` is?

**Comment by yurrriq** [comment]
LGTM, though I haven’t tried building in a few months.

**Comment by 7c6f434c** [comment]
@GrahamcOfBorg build idrisPackages.lightyear

**Comment by 7c6f434c** [comment]
A dirty solution for non-list `IDRIS_LIBRARY_PATH` is to symlink-merge user-provided and system library paths into a temporary directory on start…

**Review by shlevy — APPROVED** [review]

**Comment by shlevy** [comment]
Ready to go?

**Comment by mpickering** [comment]
Yes, do you want me to squash the commits?

**Comment by shlevy** [comment]
Up to you, I'll merge whenever.

**Comment by yurrriq** [comment]
Thanks for this. I’m excited to use it.

**Comment by mpickering** [comment]
I rebased it now and squash most of the commits together. 

Let me know if you find any problems @yurrriq !

**Comment by yurrriq** [comment]
I want to be able to override `buildPhase` and `checkPhase`. I thought I could before. How can I now?

**Comment by shlevy** [comment]
@yurrriq In the worst case you can use `overrideDerivation` to set phases, but if this is a common need a PR to properly extend the interface would probably be better.

**Comment by yurrriq** [comment]
I personally vote for support in the PR. With that said, I'm having all kinds of weird issues with running tests and generating executables, such that this doesn't work for me. It's unclear if it's my fault or an issue with this PR, and I can't commit to a dive deep terribly soon.

**Comment by mpickering** [comment]
I can't analyse whether I have broken everything with this without an error message or any reproduction steps. I think someone should merge this and then another user might report exactly where it is wrong. The common cases work for me.

**Pull request merged**